### PR TITLE
[PTQ] Add passes for PTQ to create inference NNCFGraph

### PR DESCRIPTION
**Pull request by KodiaqQ**
### Changes

- Added pass for PTQ to create inference NNCFGraph

### Reason for changes

- Deletion of ShapeOf subgraphs will allow Solver to insert quantizers correctly

### Related tickets

- 103931

### CODE REVIEW

File: tests/openvino/native/quantization/test_sanity.py [hunk 1]
```diff
@@ -29,7 +29,7 @@
     ('resnet-18-pytorch', 'imagenette2-320', {'accuracy@top1': '0.784', 'accuracy@top5': '0.946'}),
     ('mobilenet-v3-small-1.0-224-tf', 'imagenette2-320', {'accuracy@top1': '0.733', 'accuracy@top5': '0.913'}),
     ('googlenet-v3-pytorch', 'imagenette2-320', {'accuracy@top1': '0.915', 'accuracy@top5': '0.994'}),
-    ('mobilefacedet-v1-mxnet', 'wider', {'map': '0.7769245134847838'}),
+    ('mobilefacedet-v1-mxnet', 'wider', {'map': '0.7769248380270294'}),
```
Comment (l-bat) [inline tests/openvino/native/quantization/test_sanity.py:32]: Why did you change reference metric?
I guess the test should pass with abs=0.006: `0.7769248380270294 == pytest.approx(0.7769245134847838, abs=0.006)`

**Review by l-bat — APPROVED** [review]
Thanks!

**Review by kshpv — APPROVED** [review]
The new onnx quantization graphs look correct. Thank you!

File: nncf/common/graph/operator_metatypes.py [hunk 2]
```diff
@@ -169,3 +169,14 @@ class OutputNoopMetatype(OperatorMetatype):
     @classmethod
     def get_all_aliases(cls) -> List[str]:
         return [NNCFGraphNodeType.OUTPUT_NODE]
+
+
+
+#pylint:disable=no-member
+@NOOP_METATYPES.register()
+class ShapeOfNoopMetatype(OperatorMetatype):
```
Comment (andrey-churkin) [inline nncf/common/graph/operator_metatypes.py:177]:
```suggestion
class ShapeOfMetatype(OperatorMetatype):
```

**Comment by KodiaqQ** [comment]
Jenkins retest this please

**Comment by KodiaqQ** [comment]
run tensorflow pre-commit tests

**Comment by KodiaqQ** [comment]
run pytorch pre-commit tests

File: nncf/quantization/passes.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        shape_of_queue = collections.deque()
+        shape_of_queue.extend(nncf_graph.get_next_nodes(shape_of_node))
+        while shape_of_queue:
+            node = shape_of_queue.pop()
+            if node in nodes_to_drop or node in infer_nodes:
+                continue
+            nodes_to_drop.add(node)
+            # traverse forward and backward to exclude full shape of subgraph
+            # recursion excluded due to infer_nodes list around subgraph shape
+            shape_of_queue.extend(nncf_graph.get_next_nodes(node) + nncf_graph.get_previous_nodes(node))
```
Comment (andrey-churkin) [inline nncf/quantization/passes.py:66]: Just for my understanding. Why this traverse is required? Does this implementation wrong?
```python
def remove_shape_of_subgraphs(nncf_graph: NNCFGraph) -> NNCFGraph:
    visited = {node: True for node in nncf_graph.get_input_nodes()}
    queue = collections.deque(visited)
    while queue:
        curr_node = queue.popleft()
        for v in nncf_graph.get_next_nodes(curr_node):
            if visited.get(v, False) or issubclass(v.metatype, ShapeOfMetatype):
                continue
            queue.append(v)
            visited[v] = True
    nodes_to_drop = (node for node in nncf_graph.get_all_nodes() if not visited.get(node, False))
    nncf_graph.remove_nodes_from(nodes_to_drop)

    return nncf_graph
```

File: nncf/quantization/passes.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    :param nncf_graph: NNCFGraph instance for the transformation.
+    :return: NNCFGraph without ShapeOf subgraphs.
+    """
+    nodes_to_drop = set()
+    shape_of_nodes = []
+    infer_nodes = []
+
+    nodes_queue = collections.deque()
+    nodes_queue.extend(nncf_graph.get_input_nodes())
```
Comment (andrey-churkin) [inline nncf/quantization/passes.py:43]: We should add quantizers here is we want to use this implementation in the accuracy-aware algorithm.

File: nncf/quantization/passes.py (hunk 3 above)
Comment (KodiaqQ) [reply nncf/quantization/passes.py:66]: `ShapeOf` subgraphs can be enclosed and generate their own constants that need to be gently removed to prevent the appearance of residues after the removal of the main nodes.
image
image

File: nncf/quantization/passes.py (hunk 4 above)
Comment (KodiaqQ) [reply nncf/quantization/passes.py:43]: Why should we add quantizers here? How it can be used in the accuracy-aware scenario? This method removes `ShapeOf` subgraphs that are constant, not just any nodes that we want.

Comment (andrey-churkin) [reply nncf/quantization/passes.py:43]: Does this method correctly remove ShapeOf subgraphs from a quantized model graph?

File: nncf/quantization/passes.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        nodes_queue.extend(nncf_graph.get_next_nodes(node))
+
+    for shape_of_node in shape_of_nodes:
+        nodes_to_drop.add(shape_of_node)
+
+        shape_of_queue = collections.deque()
+        shape_of_queue.extend(nncf_graph.get_next_nodes(shape_of_node))
+        while shape_of_queue:
+            node = shape_of_queue.pop()
+            if node in nodes_to_drop or node in infer_nodes:
```
Comment (andrey-churkin) [inline nncf/quantization/passes.py:61]: `node in infer_nodes` -- `O(len(infer_nodes))`

Is it possible to use `dict` instead of a `list` to check this?

File: nncf/quantization/passes.py (hunk 4 above)
Comment (KodiaqQ) [reply nncf/quantization/passes.py:43]: It should correctly remove the `ShapeOf` subgraphs from the NNCFGraph with the quantizers. But this functionality wasn't tested because the current PTQ pipeline does not include such a scenario.

Comment (andrey-churkin) [inline nncf/quantization/passes.py:43]: Minor improvements. Up to you
```suggestion
    nodes_queue = collections.deque(nncf_graph.get_input_nodes())
```

File: nncf/quantization/passes.py (hunk 5 above)
Comment (KodiaqQ) [reply nncf/quantization/passes.py:61]: It's possible. But I don't see any significant performance improvements from this.

Comment (andrey-churkin) [reply nncf/quantization/passes.py:61]: ok

**Review by andrey-churkin — APPROVED** [review]

**Comment by KodiaqQ** [comment]
run tensorflow pre-commit tests

File: nncf/common/graph/operator_metatypes.py [hunk 6]
```diff
@@ -169,3 +169,14 @@ class OutputNoopMetatype(OperatorMetatype):
     @classmethod
     def get_all_aliases(cls) -> List[str]:
         return [NNCFGraphNodeType.OUTPUT_NODE]
+
+
+
+#pylint:disable=no-member
+@NOOP_METATYPES.register()
```
Comment (andrey-churkin) [inline nncf/common/graph/operator_metatypes.py:176]:
```suggestion
```

**Review by andrey-churkin — CHANGES_REQUESTED** [review]

**Review by vshampor — APPROVED** [review]

**Review by andrey-churkin — APPROVED** [review]
